neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,260 @@
+---
+NEP: 519
+Title: Yield Execution
+Authors: Akhi Singhania <akhi3030@gmail.com>; Saketh Are <saketh@near.org>
+Status: Draft
+DiscussionsTo: https://github.com/near/NEPs/pull/519
+Type: Protocol
+Version: 0.0.0
+Created: 2023-11-17
+LastUpdated: 2023-11-20
+---
+
+## Summary
+
+Today, when a smart contract is called by a user or another contract, it has no sensible way to delay responding to the caller till it has observed another future transaction.  This proposal introduces this possibility into the NEAR protocol.
+
+## Motivation
+
+There exist some situations where when a smart contract on NEAR is called, it will only be able to provide an answer at some arbitrary time in the future.  So the callee needs a way to defer replying to the caller till this time in future.
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract defers replying to the caller.
+3. External indexers are monitoring the transactions on the contract; they observe the new signing request, compute a signature, and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not yet ready to respond
+/// to its caller yet.  The smart contract promises to call
```

Comment:
Same note as earlier, `promise_yield_create` by itself does not imply this behaviour. It only occurs in combination with `promise_return`. Here I think the `promise_yield_create` should be documented about its own behaviour and only that, since it is a specification of these functions.

There is also no real punishment to anybody if they don't call a corresponding `promise_yield_resume`, so I would rephrase the 2nd sentence in terms of the number of blocks the protocol is willing to keep track of the receipt before timing it out and invoking the resumption function with no data.